Add CA1872: Prefer 'Convert.ToHexString' over 'BitConverter.ToString'

mpidash opened the pull request:
Fixes dotnet/runtime#81796.

This analyzer detects the usage of the pattern `BitConverter.ToString(bytes).Replace("-", "")` to convert an array of bytes to an uppercase hex string (without hyphens in between) and replaces it with a call to `Convert.ToHexString(bytes)`.
The analyzer will also try to preserve chaining `ToLower*` in between (before and after `string.Replace`) for a lowercase hex string.

I have found 2 warnings in `dotnet/runtime`, 1 warning in `dotnet/roslyn` and no warnings in `dotnet/aspnetcore` and `dotnet/roslyn-analyzers`:
  1. [runtime\src\coreclr\tools\aot\ILCompiler.Compiler\Compiler\XmlObjectDumper.cs(73,20): info CA1872: Prefer 'System.Convert.ToHexString(byte[])' over 'System.BitConverter.ToString(byte[])'](https://github.com/dotnet/runtime/blob/420717ee369d44906f5ba6bbf94ee28058c56037/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/XmlObjectDumper.cs#L73)
  2. [runtime\src\coreclr\tools\Common\Compiler\NativeAotNameMangler.cs(125,38): warning CA1872: Prefer 'System.Convert.ToHexString(byte[])' over 'System.BitConverter.ToString(byte[])'](https://github.com/dotnet/runtime/blob/420717ee369d44906f5ba6bbf94ee28058c56037/src/coreclr/tools/Common/Compiler/NativeAotNameMangler.cs#L125)
  3. [roslyn\src\Tools\BuildValidator\CompilationDiff.cs(404,46): info CA1872: Prefer 'System.Convert.ToHexString(byte[])' over 'System.BitConverter.ToString(byte[])](https://github.com/dotnet/roslyn/blob/7c457a6d793d0c88f58ed0be97df7a381b586ecd/src/Tools/BuildValidator/CompilationDiff.cs#L404)

mpidash commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferConvertToHexStringOverBitConverter.Fixer.cs:
```diff
@@ -0,0 +1,102 @@
+                // BitConverter.ToString(data).ToLower().Replace("-", "") => Convert.ToHexString(data).ToLower()
+                if (toLowerInvocation is not null)
+                {
+                    methodInvocation = generator.InvocationExpression(
```
At first, I tried to use the most fitting `ToLower*` method here instead of using the one in the original code, but I got many diagnostics:

- `Convert.ToHexString(data).ToLower()` triggers both CA1304 and CA1311
- `Convert.ToHexString(data).ToLowerInvariant()` triggers CA1308
- `Convert.ToHexString(data).ToLower(CultureInfo.CurrentCulture)` does not trigger a diagnostic, but seems like excessive when only converting hex characters.

So I decided to use the original `ToLower*` invocation here.

mpidash replied:
This can be simplified when https://github.com/dotnet/runtime/issues/60393 is ready.

buyaa-n replied:
> This can be simplified when https://github.com/dotnet/runtime/issues/60393 is ready.

This is now fixed, could you update?

mpidash replied:
I've updated the fixer to use `Convert.ToHexStringLower` if it is available (>= .NET 9 Preview 1) in 0efec42ac182d00a19c54f9665d33043689a1f5d, but I had to include a temporary `ReferenceAssemblies` in the tests for now.

We should wait with merging this PR until https://github.com/dotnet/roslyn-sdk/pull/1149 is ready.

mpidash replied:
I've removed the temporary reference assembly in f3fee4aec29fa3e688245c44f9736559fd493573.

mpidash commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferConvertToHexStringOverBitConverter.Fixer.cs:
```diff
@@ -0,0 +1,102 @@
+                    // BitConverter.ToString(data).Replace("-", "") => Convert.ToHexString(data)
+                    1 => generator.InvocationExpression(methodExpression, bitConverterArgumentsInParameterOrder[0].Value.Syntax),
+                    // BitConverter.ToString(data, start).Replace("-", "") => Convert.ToHexString(data, start, data.Length - start)
+                    2 => generator.InvocationExpression(
```
Not sure if we should be more restrictive with the `data` argument in this case, as the fixer changes the code to access it twice.

buyaa-n replied:
What about converting to span overload instead `Convert.ToHexString(data.AsSpan().Slice(start))`?

mpidash replied:
Fixed in 6db3a8483d0c7684aebd6f008441cf5834931f21.

mpidash commented on src/Utilities/Compiler/DiagnosticCategoryAndIdRanges.txt:
```diff
@@ -12,7 +12,7 @@
 Globalization: CA2101, CA1300-CA1311
 Mobility: CA1600-CA1601
-Performance: HA, CA1800-CA1870
+Performance: HA, CA1800-CA1872
```
Skipping `CA1871` because it is used by #6815.

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -195,6 +195,18 @@
+    <value>Prefer 'Convert.ToHexString' over 'BitConverter.ToString'</value>
+  </data>
+  <data name="PreferConvertToHexStringOverBitConverterDescription" xml:space="preserve">
+    <value>'Convert.ToHexString' is more efficient and allocation-friendly than 'BitConverter.ToString' in combination with 'string.Replace'.</value>
```
NIT: Probably it is important to mention that it's not producing dashes
```suggestion
    <value>'Convert.ToHexString' do not produce dashes, so no need to remove them. Plus, it is more efficient and allocation-friendly than using 'BitConverter.ToString'.</value>
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -195,6 +195,18 @@
+    <value>'Convert.ToHexString' is more efficient and allocation-friendly than 'BitConverter.ToString' in combination with 'string.Replace'.</value>
+  </data>
+  <data name="PreferConvertToHexStringOverBitConverterMessage" xml:space="preserve">
+    <value>Prefer '{0}' over '{1}'</value>
```
NIT: All diagnostics would include `.Replace("-", "")` part right?
```suggestion
    <value>Prefer '{0}' over '{1}.Replace("-", "")'</value>
```

mpidash replied:
There can also be a `ToLower` in between `BitConverter.ToString(data)` and `Replace("-", "")`, or replace could use `string.Empty` as second parameter.

I will try to add better message arguments when modifying the analyzer to also handle `Convert.ToHexStringLower`.

mpidash commented:
@buyaa-n I've addressed your feedback and also tried to improve the messages, PTAL 😸.

Here are the findings with the new message template:

```
W:\roslyn\src\Tools\BuildValidator\CompilationDiff.cs(404,46): info CA1872: Prefer 'System.Convert.ToHexString(byte[])' over call chains based on 'System.BitConverter.ToString(byte[])'

W:\runtime\src\coreclr\tools\Common\Compiler\NativeAotNameMangler.cs(111,38): warning CA1872: Prefer 'System.Convert.ToHexString(byte[])' over call chains based on 'System.BitConverter.ToString(byte[])' (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1872) [W:\runtime\src\coreclr\tools\aot\ILCompiler.Compiler\ILCompiler.Compiler.csproj]
W:\runtime\src\coreclr\tools\aot\ILCompiler.Compiler\Compiler\XmlObjectDumper.cs(73,20): warning CA1872: Prefer 'System.Convert.ToHexStringLower(byte[])' over call chains based on 'System.BitConverter.ToString(byte[])' (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1872) [W:\runtime\src\coreclr\tools\aot\ILCompiler.Compiler\ILCompiler.Compiler.csproj]
W:\runtime\src\libraries\System.Private.Xml\tests\XmlNodeReader\System.Xml.XmlNodeReader.Tests\XmlNodeReaderReadTests.cs(183,41): warning CA1872: Prefer 'System.Convert.ToHexString(byte[])' over call chains based on 'System.BitConverter.ToString(byte[])' (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1872) [W:\runtime\src\libraries\System.Private.Xml\tests\System.Private.Xml.Tests.csproj]
W:\runtime\src\libraries\System.Private.Xml\tests\XmlNodeReader\System.Xml.XmlNodeReader.Tests\XmlNodeReaderReadTests.cs(233,46): warning CA1872: Prefer 'System.Convert.ToHexString(byte[])' over call chains based on 'System.BitConverter.ToString(byte[])' (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1872) [W:\runtime\src\libraries\System.Private.Xml\tests\System.Private.Xml.Tests.csproj]
```

And an example of an applied code fix:

![image](https://github.com/dotnet/roslyn-analyzers/assets/4247811/0be141ce-c934-485d-b86b-a06bd8f937f4)
![image](https://github.com/dotnet/roslyn-analyzers/assets/4247811/8e7cffb2-ff02-4814-a84f-3f07e6b15a6a)
![image](https://github.com/dotnet/roslyn-analyzers/assets/4247811/4cfad5f8-91dc-4a01-995c-2656c22b9066)

@gewarren: It would be great if you could also take a look at the resource strings, thanks!

gewarren commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -195,6 +195,18 @@
+    <value>Prefer 'Convert.ToHexString' and 'Convert.ToHexStringLower' over call chains based on 'BitConverter.ToString'</value>
+  </data>
+  <data name="PreferConvertToHexStringOverBitConverterDescription" xml:space="preserve">
+    <value>Use 'Convert.ToHexString' or 'Convert.ToHexStringLower' when encoding bytes to a hexadecimal string representation, as they are more efficient and allocation-friendly than using 'BitConverter.ToString' in combination with 'string.Replace' to replace dashes and 'string.ToLower'.</value>
```
```suggestion
    <value>Use 'Convert.ToHexString' or 'Convert.ToHexStringLower' when encoding bytes to a hexadecimal string representation. These methods are more efficient and allocation-friendly than using 'BitConverter.ToString' in combination with 'string.Replace' to replace dashes and 'string.ToLower'.</value>
```

gewarren replied:
Is this analyzer C#-specific? Otherwise I think we should use capitalized "String" type in the message.

mpidash replied:
It is not, I've applied your suggestions in 309d4eae067e290180ab3feeef9a36cee0d240f2, thanks for the feedback! 😃

gewarren reviewed (commented):
I left some feedback on the resource strings. Thanks for tagging me!

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferConvertToHexStringOverBitConverter.cs:
```diff
@@ -0,0 +1,329 @@
+
+            /// <summary>
+            /// Attempts to get a BitConverter.ToString invocation, its appropriate replacement method and optionally a string.ToLower* invocation:
+            ///   1. Check if invocation is a string.ToLower* invocation. If so, continue the next steps with its instance.
+            ///   2. Check if invocation is a string.Replace("-", "") invocation. Abort if not. Continue the next steps with its instance.
+            ///   3. Check if invocation is a string.ToLower* invocation. If so, continue the next steps with its instance.
+            ///   4. Check if invocation is a BitConverter.ToString invocation. If not, abort.
```
I did not understood why the same step taken at row 1 and 3, first thought it was a typo, but the implementation looks like it indeed repeats the step, this makes me feel there is a repeated logic.

It looks checks `string.Replace` first, then matches `BitConvertor.ToString()`. because this will be called for each invocation this matches twice for `BitConverter.ToString(data).Replace("-", "").ToLower()`, looks there is workaround that omit one of the matches when there `string.ToLower` involved.

I think its better to match `BitConvertor.ToString()` first, then check if it has parent invocation `string.Replace`, then the optional `string.ToLower`. This could avoid extra match and I think will be more efficient, what you think?

buyaa-n replied:
This is NIT, you can keep it if you think the current solution is better.

mpidash replied:
I agree, starting with `BitConverter.ToString()` is better to avoid duplicate work and to get rid of the duplicate diagnostics workaround.

Checking for `string.ToLower*` twice is still necessary because the following cases are both valid (and can be fixed):

1. `BitConverter.ToString(data).ToLower().Replace("-", "")`
2. `BitConverter.ToString(data).Replace("-", "").ToLower()`

I've changed the order of invocation analysis and added additional comments (with examples) to better illustrate that `string.ToLower*` can be in two places in the call chain in fe1ca9a5e133895a13de4676e27e2a94cc1808da.

buyaa-n replied:
Thank you! Looks great to me

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferConvertToHexStringOverBitConverter.cs:
```diff
@@ -0,0 +1,329 @@
+                    (newValueIsConstantNullOrEmptyString || newValue.HasNullConstantValue() || IsStringEmptyField(newValue));
+            }
+
+            private bool TryGetBitConverterReplacement(IMethodSymbol method, [NotNullWhen(true)] out IMethodSymbol? replacement)
```
This method feels redundant, `_bitConverterReplacements` is a dictionary why we could not use its `TryGetValue` directly? `_bitConverterReplacements.TryGetValue(method, out replacement)`

mpidash replied:
Thanks, fixed in 03110732e85f9dceebacfa846ce8e4762ab7c2df, not sure what got into me when writing this helper 😅

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferConvertToHexStringOverBitConverter.cs:
```diff
@@ -0,0 +1,329 @@
+                return replacement is not null;
+            }
+
+            private bool TryGetBitConverterReplacementToLower(IMethodSymbol method, [NotNullWhen(true)] out IMethodSymbol? replacement)
```
Same here

mpidash replied:
See https://github.com/dotnet/roslyn-analyzers/pull/6967#discussion_r1527267929.

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferConvertToHexStringOverBitConverter.cs:
```diff
@@ -0,0 +1,329 @@
+
+                // Check if there is a valid replacement that uses Convert.ToHexStringLower or Convert.ToHexString.
+                if (toLowerInvocation is not null &&
+                    TryGetBitConverterReplacementToLower(invocation.TargetMethod, out var replacementToLower))
```
```suggestion
                    _bitConverterReplacementsToLower.TryGetValue(invocation.TargetMethod, out var replacementToLower))
```

mpidash replied:
Thanks, replaced it in 03110732e85f9dceebacfa846ce8e4762ab7c2df.

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferConvertToHexStringOverBitConverter.cs:
```diff
@@ -0,0 +1,329 @@
+
+                    return true;
+                }
+                else if (TryGetBitConverterReplacement(invocation.TargetMethod, out var replacement))
```
```suggestion
                else if (_bitConverterReplacements.TryGetValue(invocation.TargetMethod, out var replacement))
```

mpidash replied:
Thanks, replaced it in 03110732e85f9dceebacfa846ce8e4762ab7c2df.

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferConvertToHexStringOverBitConverter.Fixer.cs:
```diff
@@ -0,0 +1,102 @@
+            var operation = semanticModel.GetOperation(node, context.CancellationToken);
+
+            if (operation is not IInvocationOperation invocation ||
+                !PreferConvertToHexStringOverBitConverterAnalyzer.RequiredSymbols.TryGetSymbols(semanticModel.Compilation, out var symbols) ||
+                !symbols.TryGetBitConverterToStringInvocationAndReplacement(invocation, out var bitConverterInvocation, out var convertToHexStringMethod, out var toLowerInvocation))
```
Do not really think we need to populate all symbols here, we don't need string methods for example, and analyze the invocation again. I think these calls are inefficient for registering a fixer. 

Looks you only need to populate the `Convert` type, you can get the `bitConverterInvocation` from the invocation operation, also `toLowerInvocation` if it was there. Or maybe should pass the arguments using `ImmutableDictionary<string, string?>? properties` when reporting diagnostic in the analyer

mpidash replied:
You are right, getting all symbols again in the fixer seems a bit excessive.

I've passed the location of the `bitConverterInvocation` and `toLowerInvocation` (if present) as additional locations and added the name of the replacement method to the diagnostic properties in fe1ca9a5e133895a13de4676e27e2a94cc1808da.

buyaa-n commented:
> @buyaa-n I've addressed your feedback and also tried to improve the messages, PTAL 😸.

Thank you @mpidash, the messages and test coverage look great to me, functional testing worked well too, this time I looked into the code more and left few more feedback.

mpidash commented:
Thanks for the review @buyaa-n! I've addressed your feedback, PTAL.

buyaa-n reviewed (approved):
Thank you @mpidash, LGTM.

It's good to go after the https://github.com/dotnet/roslyn-analyzers/pull/6967#discussion_r1518362144 done
